lint/linter.py

```diff
@@ -438,159 +435,70 @@ def get_view_settings(self):
-        drive
-            sandbox
-                project 1
-                    path to file
-                        file to lint
-                project 2
-                    path to file
-                        file to lint
-                project 3
-                    path to file
-                        file to lint
-
-        ${project}:
-            full path of the project root directory
-            -> "/drive/sandbox/project 1"
+        """Replace tokens with values in settings.
 
-        ${directory}:
-            full path the current view's parent directory
-            -> "/drive/sandbox/project 1/path to file"
+        Generally, we support all variables Sublime Text provides:
+        "packages", "platform", "file", "file_path", file_name",
+        "file_base_name", "file_extension, "folder", "project", project_path",
+        "project_name", "project_base_name, "project_extension".
 
-        ${project} and ${directory} expansion are dependent on
-        having a window. Paths do not contain trailing directory separators.
+        Note that we ship a enhanced version for 'folder' if you have multiple
+        folders open in a window. See `_guess_project_path`.
 
-        ${home}: the user's $HOME directory.
-        ${sublime}: sublime text settings directory.
-        ${env:x}: the environment variable 'x'.
+        Additionally we expand user variables using `os.path.expanduser`.
+        See: https://docs.python.org/3/library/os.path.html#os.path.expanduser
 
+        And environment variables using `os.path.expandvars`.
+        See https://docs.python.org/3/library/os.path.html#os.path.expandvars
 
         """
-        def recursive_replace_value(expressions, value):
-            if isinstance(value, dict):
-                value = recursive_replace(expressions, value, nested=True)
+        def recursive_replace(vars, value):
+            if isinstance(value, Mapping):
+                return {key: recursive_replace(vars, val)
+                        for key, val in value.items()}
             elif isinstance(value, list):
-                value = [recursive_replace_value(expressions, item) for item in value]
+                return [recursive_replace(vars, item)
+                        for item in value]
             elif isinstance(value, str):
-                for exp in expressions:
-                    if isinstance(exp['value'], str):
-                        value = value.replace(exp['token'], exp['value'])
-                    else:
-                        value = exp['token'].sub(exp['value'], value)
-
-            return value
-
-        def recursive_replace(expressions, mutable_input, nested=False):
-            for key, value in mutable_input.items():
-                mutable_input[key] = recursive_replace_value(expressions, value)
-            if nested:
-                return mutable_input
+                value = os.path.expanduser(value)
+                value = os.path.expandvars(value)
```

Comment:
We *could* also just add `os.environ` to our `vars` dict, while having the ST variables override. `expand_variables` is case-sensitive so it will be *very* unlikely to have collisions with the ST variables.

This would then allow replacements such as `${XDG_CONFIG_HOME:$HOME/.config}/blub`. When moving expanduser to the end of the chain, `${XDG_CONFIG_HOME:~.config}/blub` even. None of this is possible currently.

Note that this won't support `%APPDATA%` replacements for Windows, but this wasn't possible previously either.